sklearn/preprocessing/_discretization.py

```diff
@@ -231,32 +246,72 @@ def fit(self, X, y=None, sample_weight=None):
 
         n_samples, n_features = X.shape
 
-        if sample_weight is not None and self.strategy == "uniform":
-            raise ValueError(
-                "`sample_weight` was provided but it cannot be "
-                "used with strategy='uniform'. Got strategy="
-                f"{self.strategy!r} instead."
-            )
-
         if self.subsample is not None and n_samples > self.subsample:
             # Take a subsample of `X`
+            # When resampling, it is important to subsample **with replacement** to
+            # preserve the distribution, in particular in the presence of a few data
+            # points with large weights.
             X = resample(
                 X,
-                replace=False,
+                replace=True,
                 n_samples=self.subsample,
                 random_state=self.random_state,
+                sample_weight=sample_weight,
             )
-
-        n_features = X.shape[1]
-        n_bins = self._validate_n_bins(n_features)
+            # Since we already used the weights when resampling when provided,
+            # we set them back to `None` to avoid accounting for the weights twice
+            # in subsequent operations to compute weight-aware bin edges with
+            # quantiles or k-means.
+            sample_weight = None
 
         if sample_weight is not None:
             sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
 
+        n_features = X.shape[1]
+        n_bins = self._validate_n_bins(n_features)
+
         bin_edges = np.zeros(n_features, dtype=object)
+
+        # TODO(1.9): remove and switch to quantile_method="averaged_inverted_cdf"
+        # by default.
+        quantile_method = self.quantile_method
+        if self.strategy == "quantile" and quantile_method == "warn":
+            warnings.warn(
+                "The current default behavior, quantile_method='linear', will be "
+                "changed to quantile_method='averaged_inverted_cdf' in "
+                "scikit-learn version 1.9 to naturally support sample weight "
+                "equivalence properties by default. Pass "
```

Comment:
What is "sample weight equivalence properties" - could we rephrase this (or am I the only one who doesn't know it)?

Reply:
the weight of a data point =n <=> this data point is repeated n times

It's described in the glossary https://scikit-learn.org/dev/glossary.html#term-sample_weight, but you'll notice that this section is not complete and not very precise :)

Reply:
@betatim thank you for the coments, I can imagine it's a lot of conversation to follow, as @jeremiedbb said, we conducted some tests with the gist [here](https://gist.github.com/snath-xoc/ef5dbe53ca4f02655fe99173e9ab1198). 

@jeremiedbb found that when specifying an odd case of sample (i.e., with one large outlier the resampling with repacement strategy held best in terms of sample equivalence (i.e., equivalent properties when weighting vs. repeating a sample n times, where n=0 means omitting the sample altogether).

We also had to modify the "uniform" strategy with @ogrisel's help and implement an averaged weighted quantile strategy with accompanying tests. Let me know if you need further clarification :) .

Reply:
I will open a PR to improve the glossary entry. This is an excellent place to document the expected semantics for `sample_weight` in scikit-learn.

Meanwhile, the precise definition of that equivalence is given in `check_sample_weight_equivalence_on_dense_data`. Note that this check is expected to fail for estimators which have a `fit` method whose outcome depends on the value of a `random_state` param (aka estimators with a non-deterministic `fit`).

The fact that an estimator has a deterministic fit or not, can both depend on the choice of the hyper-parameters (e.g. `subsample`, `max_features`, ...) and sometimes dataset size (e.g. `subsample=200_000` will enable random subsampling only on training sets with more than 200_000 rows).

For estimators with a stochastic fit, we have to run a more computational intensive statistical test to check if the equivalence property holds. This is currently done in external notebooks from this gist: https://gist.github.com/snath-xoc/fb28feab39403a1e66b00b5b28f1dcbf . We don't have specific plans to refactor those notebooks as a common test at this time because:

- they can be slow to run;
- the fact that an estimator pass or fail a statistical test depends on a choice of a pvalue threshold which is quite arbitrary, especially since we have to do multiple comparisons in our case and the individual subtests are dependent in non-trivial manners which makes adjusting for multiple comparison challenging.

So for now, we keep those notebooks as a guide to help us detect stochastic estimators with suspicious `sample_weight` handling. Bugs are manually confirmed by auditing their source code, starting from the worst offenders by ranking by ascending pvalues.

Reply:
> I will open a PR to improve the glossary entry. This is an excellent place to document the expected semantics for sample_weight in scikit-learn.

I opened in a PR: #30564.